pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -277,13 +276,53 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 	}
 
 	cidrStrings := make([]string, 0)
-	var northInterfaces networkv1.NorthInterfacesAnnotation
-	var additionalNodeNetworks networkv1.MultiNetworkAnnotation
 
 	if len(instance.NetworkInterfaces) == 0 || (len(instance.NetworkInterfaces) == 1 && len(instance.NetworkInterfaces[0].AliasIpRanges) == 0) {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: Node %v has no ranges from which CIDRs can be allocated", node.Name)
 	}
+
+	defer func() {
```

Comment:
we should avoid this large defer functions with a lot code, it makes the code harder to read and is prone to create subtles bug in the future, since adding anything after it will have to take all this code into consideration

Reply:
also the node object is mutated several times later

Reply:
also, this only makes sense if we know that we have modified the node object

Reply:
I have updated the description of this PR. main goal is to not worry about API patching and when it has to be done. This way, it is in one place, and I have to just modify the `node` variable if I need to perform any node changes. This gives us as well gating on when to even reach out to the API.

Reply:
I saw other places with this pattern, I personally think that this is an antipattern in go, it may be used from some other language but honestly I don't know. It makes the function less readable and I don't see the benefit of " worry about API patching a" because you have to do it always before returning. 
With current implementation , I can ready top to bottom and understand the logic. With this change I have to read top to bottom, memorize where the defer is, and remember that everything in between the end of the function and the defer has to be aware that is still code that will run

Reply:
Agree on that if we keep the current form, i.e. we do nothing when return error. But as soon as I add logic to update something when error is returned, I have to make sure non of the other "error returns" prevent me from updating the API. So then I have to go over some additional hoops to ensure I update the API, and usually this will be a 1 off, then another, which leads to having update called all over the place. With this, approach it is "cleaner" in terms of: I have an update function already defined, now I can do whatever to the object, or return however or whenever I want.

Reply:
I'm not going to push more back on this because you are the expert here, I probably miss some of those details that you are mentioning, I trust your judgment Maciej 👍 😄

Reply:
I agree with Antonio here that this is really difficult to read, reason about and maintain. If someone changes something later and does not expect the node to be updated, this function is going to do it anyways.

I think in this case, you are much better off using the 'goto' statement.

You could also split up this defer into two different ones:
defer updatePodCIDRs()
defer updateNodeAnnotations()
Then you can call them at the appropriate times.

Thus far the changes we have made for multi networking haven't been active in the GKE fleet. I am also concerned that this change will affect almost all GKE clusters and we don't have any kind of experiment guard on it.

Reply:
That is the whole point: to reduce all API changes to just the node object changes, instead of passing around different variables, I have updated description of 'updateCIDRAllocation' for that purpose.

Reply:
I have dropped the defer, but still maintain the centralized approach of modifying node object, PTAL.